Replace git cloner with go getter to support various target

yujunz opened the pull request:
Fix #923 #1272

Liujingfang1 commented:
Originally we used go-getter to support the remote targets. When we integrated kustomize into kubectl, the go-getter and it dependencies blocked the integration. So we removed it. There is no plan for adding it back for now. You can use the go-getter plugin instead. Thank you.

yujunz commented:
TIL

Shall we consider making the cloner configurable? i.e. instead of hardcoding it as `ClonerUsingGitExec`, allow executing arbitrary binary configured by user. This would not introduce any dependency on kustomize and is backward compatible.

yujunz commented:
cc @howardjohn for visibility

pwittrock commented:
> Shall we consider making the cloner configurable? i.e. instead of hardcoding it as ClonerUsingGitExec, allow executing arbitrary binary configured by user. This would not introduce any dependency on kustomize and is backward compatible.

@yujunz I think this is what @Liujingfang1 was suggesting by using a generator plugin.

pwittrock commented:
Using go-getter would be nice, we should revisit if this is possible.  IIRC, the go-getter code pulled in a massive transitive dependency list.  There may be a way to prune this through an upstream contribution.  Or perhaps it won't be as controversial now.

yujunz commented:
> @yujunz I think this is what @Liujingfang1 was suggesting by using a generator plugin.

I did implement a [generator plugin for go-getter](https://github.com/kubernetes-sigs/kustomize/tree/master/plugin/someteam.example.com/v1/gogetter). It was not so easy to configure and use.

The most user friendly way to solve the issue would be include `go-getter` package and solve the dependency issue in integration. How do we check it without merging this PR?  @Liujingfang1 

Customizable cloner still requires user to install the cloner (git, go-getter or anything) by themselves. A tradeoff for flexibility. If it looks good for all, I can make some time to implement it.

pwittrock commented:
@yujunz I'm not sure I understand the proposal?  Is it to create a new pluggable-cloner mechanism that would exec out to a cloner?  Or something else?

Liujingfang1 commented:
> The most user friendly way to solve the issue would be include go-getter package and solve the dependency issue in integration. How do we check it without merging this PR? @Liujingfang1

You can take a look at the [api/go.sum](https://github.com/kubernetes-sigs/kustomize/blob/master/api/go.sum)  to check what packages have been added when you add a new dependency. For example, `github.com/aws/aws-sdk-go` is added.

yujunz commented:
https://github.com/kubernetes-sigs/kustomize/blob/5707962df574d6b75788eb8f7408f492aaa1e1e5/api/internal/git/cloner.go#L34

The existing cloner is a simply exec of gitProgram. So if we can provide a way to make it declarative then it should work with arbitrary binary. e.g. in `kustomizeconfig.yaml`

```
cloners:
- go-getter
- aws s3 cp -r
```

Then we shall try exec `${cloner} ${resource} ${dest}` until success. @pwittrock

yujunz commented:
> You can take a look at the [api/go.sum](https://github.com/kubernetes-sigs/kustomize/blob/master/api/go.sum) to check what packages have been added when you add a new dependency. For example, `github.com/aws/aws-sdk-go` is added.

IIUC, we want to be cloud agnostic. I think the most wanted source is https archive. The go standard package should be sufficient.

I can try to tail go-getter to make it embeddable in kustomize. Does that work for you?  @Liujingfang1

Liujingfang1 commented:
> > You can take a look at the api/go.sum to check what packages have been added when you add a new dependency. For example,  github.com/aws/aws-sdk-go is added.

> IIUC, we want to be cloud agnostic. I think the most wanted source is https archive. The go standard package should be sufficient.
> 
> I can try to tail go-getter to make it embeddable in kustomize. Does that work for you? @Liujingfang1

It would be great if you can do it by go standard packages.

yujunz commented:
@Liujingfang1 I made a try to embed the go-getter package.

The changes to original go-getter are

* removed s3 and gcs from detect and get
* replaced module path of `helper/url`

Some additional notes

* don't be scared about the 180 files, most of them are test data as it is
* it still introduces several depedency
  * they are all small helper module and unlikely break upstream integration
  * we may rewrite some of them with standard package, but it may not worth doing so

Let me know if it works for you. cc @pwittrock

Liujingfang1 commented on api/go.mod:
```diff
@@ -16,3 +23,5 @@ require (
 	sigs.k8s.io/yaml v1.1.0
 )
+
+replace sigs.k8s.io/kustomize/api/internal/getter/helper/url => ./helper/url
```
Why adding this `replace`? I don't think it's needed.

yujunz replied:
removed

Liujingfang1 commented on api/loader/loader.go:
```diff
@@ -19,16 +26,43 @@ import (
-		return newLoaderAtGitClone(
-			repoSpec, fSys, nil, git.ClonerUsingGitExec)
+
+	root, errD := demandDirectoryRoot(fSys, target)
```
ditto

yujunz replied:
ditto

Liujingfang1 commented on api/loader/fileloader.go:
```diff
@@ -164,30 +164,49 @@ func (fl *fileLoader) New(path string) (ifc.Loader, error) {
-	if err == nil {
-		// Treat this as git repo clone request.
+
+	root, errD := func() (filesys.ConfirmedDir, error) {
```
The previous order is first to treat target as a remote target; when failed, treat it as a local target. The changes make the order reverted. Is there any reason for doing this?

yujunz replied:
In fact, getter is capable to handle local target so we cannot check it with the error return as `repoSpec, err := git.NewRepoSpecFromUrl(path)`. However it may not behavior exactly the same, e.g. the containment violation. 

Putting local target before getter is to ensure backward compatibility.

yujunz commented:
> Among the 180 files, lots of them are not needed. Can you do a clean up?
I can try my best to clean up a bit.

> I'm also wondering if you can push this go-getter version to another branch of the upstream go-getter.

Not sure it can be accepted but we definitely can propose that.

Any recommended justification?

yujunz commented:
Pushed an update with dependency on a forked go-getter.

It is quite tricky to handle `go mod` with branches or forks. `replace` does not work well. So I ended up in modifying the package name in code. See https://github.com/yujunz/go-getter/commit/fc7e8cd6fd451576893995bd09ea616181e732cc

Do you think we can host a fork under kubernetes-sigs organization? That would make the patch in kustomize cleaner and easier to maintain.  @Liujingfang1 @monopole

yujunz commented:
Link upstream discussion about reducing dependency: https://github.com/hashicorp/go-getter/issues/193

Liujingfang1 commented:
Let's merge this for now. When we can get some updates from go-getter side for  hashicorp/go-getter#193, we can change the dependency to the lite version of go-getter. Thank you for working on this. @yujunz  As a follow up, can you add an integration test or an example test for this? 

/lgtm
/approve
